### ISSUE TITLE
Undefined behavior from `poll_fn` + `futures::join!` / `std::future::join!`

### ISSUE THREAD

**Issue by jplatte** [rust-lang/miri#3780]
Recently, when working on a library for observables, I found new miri test failures (I had been getting them on a specific test before, but didn't investigate after ensuring it wasn't my own UB). This lead to a [bug report to `tokio`](https://github.com/tokio-rs/tokio/issues/6729) where Alice Ryhl showed that it didn't actually require `tokio::spawn` (which I had failed to get rid of in minimizing the issue), and the same issue was reproducible with `join!` + `futures_executor`.

I have now reduced the example much further, to this (works with the nightly `std::future::join!`, or `futures::join!`; [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=b3e2e8314b2076147edab54938547c7a)):

```rust
#![feature(future_join)]

use std::{
    future::{join, poll_fn, Future},
    pin::pin,
    sync::Arc,
    task::{Context, Poll, Wake},
};

fn main() {
    let mut fut = pin!(async {
        let mut thing = "hello".to_owned();
        let fut1 = async move {
            poll_fn(|_| {
                thing += ", world";
                Poll::<()>::Pending
            })
            .await;
        };

        let fut2 = async {};

        join!(fut1, fut2).await;
    });

    let waker = Arc::new(DummyWaker).into();
    let mut ctx = Context::from_waker(&waker);
    _ = fut.as_mut().poll(&mut ctx);
    _ = fut.as_mut().poll(&mut ctx);
}

struct DummyWaker;

impl Wake for DummyWaker {
    fn wake(self: Arc<Self>) {}
}
```

Alice suggested that this may be an instance of https://github.com/rust-lang/rust/issues/63818, is that correct? I was under the impression that the compiler for now doesn't implement alias restrictions for `!Unpin` types, but maybe this only applies to the LLVM backend, not miri?

**Comment by Darksonn** [comment]
In the original bug report to Tokio, the UB was happening even though `tokio::spawn` was used instead of `join!`. So it may not be the `join!` macro's fault.

**Comment by taiki-e** [comment]
It seems adding `move` to `poll_fn` silences SB violation error: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=06750122ba5cafa878956f1a49cd4177)

**Comment by taiki-e** [comment]
> adding `move` to `poll_fn` silences SB violation error

Or making `thing` `&mut _`: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=d37e72627b8d7cb7500f152756dd4da4)
(This reminds me of https://internals.rust-lang.org/t/surprising-soundness-trouble-around-pollfn/17484)

Btw, it is also possible to reproduce the error with a function, such as `futures::future::join()`, rather than a macro: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=fac4795c2a2f4e97cceff3e0d59ed9f9)

**Comment by oli-obk** [comment]
Minimized:

```rust
use std::{
    future::{poll_fn, Future},
    pin::Pin,
    sync::Arc,
    task::{Context, Poll, Wake},
};

fn main() {
    let mut thing = "hello".to_owned();
    let fut = async move {
        poll_fn(|_| {
            thing += ", world";
            Poll::<()>::Pending
        })
        .await;
    };

    let mut fut = MaybeDone::Future(fut);
    let mut fut = unsafe { Pin::new_unchecked(&mut fut) };

    let waker = Arc::new(DummyWaker).into();
    let mut ctx = Context::from_waker(&waker);
    assert_eq!(fut.as_mut().poll(&mut ctx), Poll::Pending);
    assert_eq!(fut.as_mut().poll(&mut ctx), Poll::Pending);
}

struct DummyWaker;

impl Wake for DummyWaker {
    fn wake(self: Arc<Self>) {}
}

pub enum MaybeDone<F: Future> {
    Future(F),
    Done,
}
impl<F: Future<Output = ()>> Future for MaybeDone<F> {
    type Output = ();

    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        unsafe {
            match *self.as_mut().get_unchecked_mut() {
                MaybeDone::Future(ref mut f) => Pin::new_unchecked(f).poll(cx),
                MaybeDone::Done => unreachable!(),
            }
        }
    }
}
```

Removing the `MaybeDone::Done` variant will also hide the UB

**Comment by RalfJung** [comment]
The memory that the conflict occurs on is the `fut` itself. Do I understand correctly that `async move` means we are moving `thing` into the future, and then creating a reference to it? So the future has (at least) two fields, a `String` and a `ThingAdder` pointing to that string.

The `match *self.as_mut().get_unchecked_mut()` will read the discriminant of the `MaybeDone`, which has been niche-optimized to be stored inside the `String. This read overlaps with where the `thing` points to, thus invalidating the mutable reference. Finally, the mutable reference is used again, causing UB.

In other words, this is the "mutable reference aliasing" version of why `UnsafeCell` blocks niches: we also need to block niches for fields that have self-referential references pointing to them. So the fix here is to change async lowering to use [`UnsafePinned`](https://github.com/rust-lang/rust/issues/125735) and then have `UnsafePinned` block niches the same way `UnsafeCell` blocks niches.

**Comment by RalfJung** [comment]
Since Miri is behaving as intended here, I'm going to close the issue in favor of https://github.com/rust-lang/rust/issues/63818 and https://github.com/rust-lang/rust/issues/125735. I don't see a way that we can work around this akin to the `!Unpin` hack; this fundamentally needs the layout of these futures to be different from what it is today.

Thanks for bringing this up! In retrospect it is somewhat surprising that it took us so long to discover this, given that we already knew about the exact same issue with shared references and `UnsafeCell`...

Cc @rust-lang/opsem, just an FYI :)

**Comment by jplatte** [comment]
> In retrospect it is somewhat surprising that it took us so long to discover this

I'm just as surprised! 😄

Given that the `&mut` method I'm trying to call in the real code that made me look into this is on my own type (not `String`), is there something I can do hide its niches without using `unsafe`, as a temporary measure to unblock me from using miri to detect other potential unsoundness?

**Comment by RalfJung** [comment]
If you only need `&mut` access, you could wrap the thing the mutable reference points to in an `UnsafeCell`. :joy:

**Comment by Darksonn** [comment]
Is there anything Tokio should be doing to avoid UB with niche-optimized enums? We have the [`MaybeDone`](https://github.com/tokio-rs/tokio/blob/master/tokio/src/future/maybe_done.rs) enum:
```Rust
pin_project! {
    /// A future that may have completed.
    #[derive(Debug)]
    #[project = MaybeDoneProj]
    #[project_replace = MaybeDoneProjReplace]
    pub enum MaybeDone<Fut: Future> {
        /// A not-yet-completed future.
        Future { #[pin] future: Fut },
        /// The output of the completed future.
        Done { output: Fut::Output },
        /// The empty variant after the result of a [`MaybeDone`] has been
        /// taken using the [`take_output`](MaybeDone::take_output) method.
        Gone,
    }
}
```

**Comment by RalfJung** [comment]
I don't know if you *should* be doing anything, since it seems unlikely that this will lead to actual miscompilations. Though maybe @comex is able to trick LLVM into generating bad code here again. ;)

What you *could* do it mark that enum with a `repr`, which AFAIK prevents it being being niche-optimized.

The best thing to do would be to find someone who can implement `UnsafePinned`. :)

**Comment by taiki-e** [comment]
> What you _could_ do it mark that enum with a `repr`, which AFAIK prevents it from being niche-optimized.

While this is adequate as a workaround for the specific case here, it does not appear to be a reasonable workaround for this problem as a whole. This is because the same issue can be caused by using Option instead of a custom enum like MaybeDone.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=63ab1f219beadfb593aac4ec5e0952d7

**Comment by RalfJung** [comment]
Yeah I did not mean to imply that this is a reasonable work-around for this case, I just pointed out the possibility.

I don't think there is a good work-around, this fundamentally needs rustc layout computation to change -- either by special-casing generators, or by implementing https://github.com/rust-lang/rust/issues/125735.

**Comment by taiki-e** [comment]
> I don't think there is a good work-around, this fundamentally needs rustc layout computation to change -- either by special-casing generators, or by implementing [rust-lang/rust#125735](https://github.com/rust-lang/rust/issues/125735).

I thought wrapping a coroutine with a MaybeUninit inside a future returned by an async block/async function (like https://github.com/crossbeam-rs/crossbeam/pull/834) as a potential workaround.

In previous async lowing I think it was enough to change the [`GenFuture`](https://github.com/rust-lang/rust/blob/1.66.0/library/core/src/future/mod.rs#L81) field from `T` to `MaybeUninit<T>`, but I'm not sure what should be done since such a type is not used in recent async lowing.

**Comment by RalfJung** [comment]
Here's a proper work-around: https://github.com/rust-lang/rust/pull/129313. It increases the size of some futures, though...
